shm/Store.cpp

```diff
@@ -1,33 +1,170 @@
 #include "Store.hpp"
 
 #include <algorithm>
+#include <iomanip>
+#include <random>
 
-Store::Store(std::shared_ptr<Time>& time)
-    : time_(time) {
-    time_->addObserver(this);
+#include "Alcohol.hpp"
+#include "Fruit.hpp"
+#include "Item.hpp"
+
+constexpr size_t rarityAmount{4};
+constexpr Rarity possibleRarities[rarityAmount]{Rarity::common, Rarity::rare, Rarity::epic, Rarity::legendary};
+
+Store::Store(std::shared_ptr<Time>& time) {
+    time_ = time;
+    time->addObserver(this);
+    makeStock();
+}
+
+size_t Store::genRand(int min, int max) const {
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<> distirb(min, max);
+
+    return static_cast<size_t>(distirb(gen));
+}
+
+void Store::generateFruits() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Fruit fruit(fruitNames[genRand(0, 5)], genRand(1, 20), genRand(10, 30), genRand(1, 10));
+        if(std::none_of(begin(market_), end(market_),[&fruit](const auto& ptr){ return ptr->getName() == fruit.getName();})) {
+            market_.emplace_back(std::make_shared<Fruit>(fruit));
+            i++;
+        }
+    }
+}
+
+void Store::generateAlcos() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Alcohol alco(alcoNames[genRand(0, 5)], genRand(1, 5), genRand(30, 100), genRand(40, 96));
+        if(std::none_of(begin(market_), end(market_),[&alco](const auto& ptr){ return ptr->getName() == alco.getName();})) {
+            market_.emplace_back(std::make_shared<Alcohol>(alco));
+            i++;
+        }
+    }
+}
+
+void Store::generateItems() {
+    size_t i = 0;
+    while(i < marketSection) {
+        Item item(itemNames[genRand(0, 5)], genRand(1, 10), genRand(30, 100), possibleRarities[genRand(0, 3)]);
+        if(std::none_of(begin(market_), end(market_),[&item](const auto& ptr){ return ptr->getName() == item.getName();})) {
+            market_.emplace_back(std::make_shared<Item>(item));
+            i++;
+        }
+    }
+}
+
+void Store::makeStock() {
+    market_.reserve(marketSection * 3);
+    generateFruits();
+    generateAlcos();
+    generateItems();
+    market_.shrink_to_fit();
```

Comment:
Is it necessary to shrink it?

Reply:
On one of the lessons there was metioned that we use this when we reserve memory, it doesn't change anything but I left it just in case